pkg/network/ovn_kubernetes.go

```diff
@@ -1604,11 +1934,130 @@ func shouldUpdateOVNKonUpgrade(ovn bootstrap.OVNBootstrapResult, releaseVersion
 		return true, true
 	}
 
-	klog.Warningf("OVN-Kubernetes daemonset versions inconsistent. node: %s, master: %s, release: %s",
+	klog.Warningf("OVN-Kubernetes daemonset versions inconsistent. node: %s, master/control-plane: %s, release: %s",
 		nodeVersion, masterVersion, releaseVersion)
 	return true, true
 }
 
+func getProgressingState(ovn bootstrap.OVNBootstrapResult) string {
+	var node, master, controlPlane string
+
+	node = "nil"
+	if ovn.NodeUpdateStatus != nil {
+		node = fmt.Sprintf("%t", ovn.NodeUpdateStatus.Progressing)
+	}
+
+	master = "nil"
+	if ovn.MasterUpdateStatus != nil {
+		master = fmt.Sprintf("%t", ovn.MasterUpdateStatus.Progressing)
+	}
+
+	controlPlane = "nil"
+	if ovn.ControlPlaneUpdateStatus != nil {
+		controlPlane = fmt.Sprintf("%t", ovn.ControlPlaneUpdateStatus.Progressing)
+	}
+
+	return fmt.Sprintf("progressing: ovnkube-node=%s, ovnkube-master=%s, ovnkube-control-plane=%s",
+		node, master, controlPlane)
+}
+
+// shouldUpdateOVNKonInterConnectZoneModeChange determines if we should roll out changes to
+// the node daemonset (single zone, multizone), to the master daemonset (single zone only) and
+// to the control plane deployment (multizone only), when the interconnect zone mode changes.
+//
+// When switching from single zone to multizone, we start with:
+// - single-zone node DS
+// - (single-zone) master DS
+// We then go through an intermediate step with:
+// - multizone node DS
+// - (multizone) control plane deployment
+// - (single-zone) master DS
+// This allows us to always have an instance of cluster manager running throughout the zone mode change.
+// We reach this intermediate step by first rolling out the multizone node DS and then (at the same time)
+// the control plane deployment and master DS.
+// Once all three components have rolled out, we simply remove the old (single-zone) master DS.
+//
+// For multizone to single zone, which is not officially supported, we do the same steps backwards.
+//
+// The whole procedure allows us to always have a working deployed ovnk while changing zone mode.
+//
+// To sum up:
+// - single zone -> multizone:   first roll out node,   then master+control plane; finally, remove master.
+// - multizone   -> single zone: first roll out master + control plane, then node; finally, remove control plane.
+func shouldUpdateOVNKonInterConnectZoneModeChange(ovn bootstrap.OVNBootstrapResult, targetZoneMode InterConnectZoneMode) (updateNode, updateMaster, addControlPlane bool) {
+
+	if ovn.NodeUpdateStatus == nil || ovn.MasterUpdateStatus == nil && ovn.ControlPlaneUpdateStatus == nil {
+		// Fresh cluster - full steam ahead!
+		return true, true, true
+	}
+
+	// When both DSs are in 4.13, we're in phase 1 of the upgrade from a non-IC version;
+	// phase 1 is carried out by shouldUpdateOVNKonUpgrade. Nothing to do here.
+	if !ovn.NodeUpdateStatus.InterConnectEnabled ||
+		ovn.MasterUpdateStatus != nil && !ovn.MasterUpdateStatus.InterConnectEnabled {
```

Comment:
dont need to check of ovn.MasterUpdateStatus is nil here, you already checked it on line 1989

Reply:
it's a bit more subtle than that. What you're saying would be true if on line 1989 I had `||` between all three conditions. Instead I check if node is nil or master _and_ control plane are both nil:
```
	if ovn.NodeUpdateStatus == nil || ovn.MasterUpdateStatus == nil && ovn.ControlPlaneUpdateStatus == nil {
		// Fresh cluster - full steam ahead!
		return true, true, true
	}
```
This means that we have either no status for ovnk node at all yet, or no status for both master and control-plane yet, which happens at very start. :) 

Here instead  (line 1996) I'm checking whether we're in phase 1, with ovnk node running without `--enable-interconnect` or master also without `--enable-interconnect`.